### PR TITLE
IEI-169967 Provide support for hash-based optimistic locking merge in the storage layer

### PR DESCRIPTION
**Pull request by msiemczyk**
- Removed optimistic locking merge logic (whole persistNode method) from HashBasedOptimisticLockingConfiguration as this is now pushed to storage layer when versions don't match. Updated olock generation code to ignore storage version in hash calculations.
- Modified OLockHashCalcFilter to support multiple keys (config value) to ignore in order to be able to ignore storage version.
- Improved exception handling by renaming OptimisticLockException to OLockMergeException (internal cause exception) and adding new user facing (friendly) exception ConfigurationConcurrentModificationException.
- Fixed bunch of Sonar issues.
- Fixed existing tests and added new ones for new/changed code.

### CODE REVIEW

File: dcm4che-conf/dcm4che-conf-core/src/main/java/org/dcm4che3/conf/core/olock/HashBasedOptimisticLockingConfigurationDecorator.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            ConfigNodeTraverser.traverseMapNode(configurationNode, new OLockHashCalcFilter(Configuration.VERSION_KEY));
+        }
+
+        return configurationNode;
+    }
+
+    /**
+     * Inserts #hash props into node according to @ConfigurableProperty annotations
+     */
+    public static final class HashMarkingTypesafeNodeFilter implements ConfigNodeTypesafeFilter {
```
Comment (homerocda) [inline dcm4che-conf/dcm4che-conf-core/src/main/java/org/dcm4che3/conf/core/olock/HashBasedOptimisticLockingConfigurationDecorator.java:70]: Typo? "TypeSafe" vs "Typesafe"

File: dcm4che-conf/dcm4che-conf-core/src/main/java/org/dcm4che3/conf/core/olock/OLockHashCalcFilter.java [hunk 2]
```diff
@@ -42,35 +42,58 @@
...
+     * 
+     * @param ignoredKeys One or more config keys to ignore.
+     */
+    public OLockHashCalcFilter(String... ignoredKeys) {
+        
         hashCalculationNodeStack.push(new byte[20]);
-        this.ignoredKey = ignoredKey;
+        
+        if (ignoredKeys != null) {
+            this.getIgnoredKeys().addAll(Arrays.asList(ignoredKeys));
```
Comment (homerocda) [inline dcm4che-conf/dcm4che-conf-core/src/main/java/org/dcm4che3/conf/core/olock/OLockHashCalcFilter.java:90]: Suggestion:

`this.ignoredKeys = Arrays.asList(requireNonNull(ignoredKeys));`

**Review by homerocda — APPROVED** [review]

**Review by hczedik — APPROVED** [review]

**Comment by hczedik** [comment]
Approved, but I haven't reviewed it in detail, because the changes are only in conf classes that are anyway not used by the client.

**Review by greg-knox — APPROVED** [review]

File: dcm4che-conf/dcm4che-conf-core/src/main/java/org/dcm4che3/conf/core/olock/HashBasedOptimisticLockingConfigurationDecorator.java (hunk 1 above)
Comment (msiemczyk) [reply dcm4che-conf/dcm4che-conf-core/src/main/java/org/dcm4che3/conf/core/olock/HashBasedOptimisticLockingConfigurationDecorator.java:70]: It was like that before, but I can rename if you like. Let me know.

**Review by homerocda — APPROVED** [review]

**Pull request merged**